server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1454,7 +1446,15 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                         try {
                             listener.onResponse(new UpdateIndexShardSnapshotStatusResponse());
                         } finally {
-                            endCompletedSnapshots(newState);
```

Comment:
This saves a ton of cycles. Just think about a 10k shards snapshot. Worst case that would have triggered ~10k ^ 2 checks on whether all shards have completed if the last snapshot update ...